sdk/nodejs/serverless/function.ts

```diff
@@ -0,0 +1,184 @@
+            closure, serializedFileNameNoExtension, options.excludePackages);
+
+        this.bucket = new storage.Bucket(`${name}-bucket`, {
+            project: options.project
+        }, { parent: this });
+
+        this.bucketObject = new storage.BucketObject(`${name}-bucket-object`, {
+            bucket: this.bucket.name,
+            source: new pulumi.asset.AssetArchive(codePaths),
+        }, { parent: this });
+
+        // Create the Cloud Function.
+        this.function = new cloudfunctions.Function(name, {
+            entryPoint: handlerName,
+            sourceArchiveBucket: this.bucket.name,
+            sourceArchiveObject: this.bucketObject.name,
+            triggerHttp: true,
+
+            availableMemoryMb: options.availableMemoryMb,
+            description: options.description,
+            labels: options.labels,
+            project: options.project,
+            region: options.region,
+            timeout: options.timeout
+        }, { parent: this });
+    }
+}
+
+// computeCodePaths calculates an AssetMap of files to include in the Function package.
+async function computeCodePaths(
+        closure: Promise<pulumi.runtime.SerializedFunction>,
+        serializedFileNameNoExtension: string,
+        excludePackages?: string[]): Promise<pulumi.asset.AssetMap> {
+
+    const serializedFunction = await closure;
+
+    excludePackages = excludePackages || [];
+
+    const excludedPackages = new Set<string>();
+    for (const p of excludePackages) {
+        excludedPackages.add(p);
+    }
+
+    // Include packages to package.json file
+    const packageJson = await producePackageJson(excludedPackages);
+
+    return {
+        // Always include the serialized function.
+        [serializedFileNameNoExtension + ".js"]: new pulumi.asset.StringAsset(serializedFunction.text),
+        // Include package.json and GCP will install packages for us
+        ["package.json"]: new pulumi.asset.StringAsset(packageJson)
+    };
+}
+
+// Get the list of packages declare in package.json, exclude pulumi and exclusions made by user,
+// generate the new package.json to be uploaded to GCP.
+// GCP will restore the packages itself.
+function producePackageJson(excludedPackages: Set<string>): Promise<string> {
+    return new Promise((resolve, reject) => {
+        readPackageJson(filepath.basename('package.json'), null, false, (err, data) => {
+            if (err) {
+              return reject(err);
+            }
+
+            const filteredDependencies = Object.keys(data.dependencies)
+                .filter(pkg => !excludedPackages.has(pkg) && !pkg.startsWith("@pulumi"))
+                .reduce((obj, key) => {
+                    obj[key] = data.dependencies[key];
+                    return obj;
+                }, {});
+
+            const newPackageJson = {
+                name: data.name,
+                description: data.description,
+                dependencies: filteredDependencies
+            }
```

Comment:
I don't actually know exactly what GCF does to install packages, but is it possible it needs more sections of the `package.json` to be fully correct?  Like `scripts` for install hooks?  Is it worth keeping everything by default and just overwriting `dependencies` field?